app/display/runtime/src/main/java/org/csstudio/display/builder/runtime/WidgetRuntime.java

```diff
@@ -422,24 +445,66 @@ public void writePV(final String pv_name, final Object value) throws Exception
                 name_to_check = name_to_check.substring(0, sep);
         }
         awaitStartup();
-        final List<RuntimePV> safe_pvs = writable_pvs;
+        List<RuntimePV> safe_pvs = writable_pvs;
         if (safe_pvs != null)
             for (final RuntimePV pv : safe_pvs)
                 if (pv.getName().equals(name_to_check))
                 {
-                    try
-                    {
-                        pv.write(value);
-                    }
-                    catch (final Exception ex)
-                    {
-                        throw new Exception("Failed to write " + value + " to PV " + name_to_check, ex);
-                    }
-                    return;
+                    return pv;
                 }
         throw new Exception("Unknown PV '" + pv_name + "' (expanded: '" + name_to_check + "')");
     }
 
+    /** Write a value to a PV
+     *  @param pv_name Name of PV to write, may contain macros
+     *  @param value Value to write
+     *  @throws Exception on error
+     */
+    public void writePV(final String pv_name, final Object value) throws Exception
+    {
+        RuntimePV pv = getWritablePV(pv_name);
+        try {
+            pv.write(value);
+        } catch (final Exception ex) {
+            throw new Exception("Failed to write " + value + " to PV " + pv_name, ex);
+        }
+    }
+
+    public void callPV(final String pv_name, final HashMap<String, String> args, final String return_pv) throws Exception {
+        final String expanded = MacroHandler.replace(widget.getMacrosOrProperties(), pv_name);
+        awaitStartup();
+        try (PVAClient client = new PVAClient()) {
+            List<PVAData> params = new ArrayList<>();
+            HashMap<String, RuntimePV> pvs = new HashMap<>();
+            getPVs().forEach(x -> pvs.put(x.getName(), x));
+            for (Map.Entry<String, String> a: args.entrySet()) {
+                String pvName = MacroHandler.replace(widget.getMacrosOrProperties(), a.getValue());
+                RuntimePV pv = pvs.get(pvName);
+                VType value = pv.read();
+
+                String argName = MacroHandler.replace(widget.getMacrosOrProperties(), a.getKey());
+
+                PVAData param;
+                // TODO: is there a generic way to convert a VType to PVAData?

```

Comment:
Is there a good way to do this? I wanted to add a constructor to (e.g.) PVAData but that would create a cross-cutting dependency, so I didn't want to just blindly go into that.